feat: Add logging util and CLI params for logging output

mkmeral opened the pull request:
## Description
This PR adds logging system to the Strands CLI. The implementation includes:

1. Command-line arguments for configuring logging:
   - `--log-level`: Set logging level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
   - `--log-file`: Specify output file path for logs

2. A new utility module `logging_utils.py` with functions for:
   - Configuring logging with file output
   - Getting available log levels
   - Checking logging status
   - Setting log levels for specific modules

3. Integration into the main CLI entry point to initialize logging based on user preferences

4. Comprehensive unit tests in `tests/utils/test_logging_utils.py` covering:
   - Log level validation
   - File handler creation
   - Directory creation
   - Exception handling
   - Logger state management
   - All edge cases and error conditions

This enhancement will help users debug issues, track agent behavior, and maintain operation logs for production deployments.

## Related Issues
[No specific issue reference provided - please add if applicable]

## Documentation PR
[Link to associated documentation PR when created]

## Type of Change
- New feature

## Testing
The changes have been thoroughly tested with:

* `hatch fmt --linter` - All linting checks pass
* `hatch fmt --formatter` - Code formatting is consistent
* `hatch test --all` - Full test suite passes

Additionally, manual testing with various logging configurations confirms the functionality:
```bash
# Test with default logging (INFO level)
strands --log-file=strands.log

# Test with debug level
strands --log-level=DEBUG --log-file=debug.log

# Verify log file is created with appropriate content
cat debug.log
```

See example
```
2025-05-25 11:06:53,210 - root - INFO - Logging configured with level: DEBUG
2025-05-25 11:06:53,210 - root - INFO - Log file: /Users/murmeral/.strands/logs/2025-05-25_11-06-52-DEBUG.log
2025-05-25 11:06:53,210 - root - INFO - Strands CLI started with log level DEBUG
2025-05-25 11:06:53,210 - root - INFO - Log file: /Users/murmeral/.strands/logs/2025-05-25_11-06-52-DEBUG.log
2025-05-25 11:06:53,210 - strands.models.bedrock - DEBUG - config=<{'model_id': 'us.anthropic.claude-3-7-sonnet-20250219-v1:0', 'max_tokens': 64000, 'additional_request_fields': {'thinking': {'type': 'enabled', 'budget_tokens': 2048}}}> | initializing
2025-05-25 11:06:53,211 - botocore.hooks - DEBUG - Changing event name from creating-client-class.iot-data to creating-client-class.iot-data-plane
...
```

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [ ] I have added an appropriate example to the documentation to outline the feature
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice

Commit:
feat: Add logging utils

Commit:
Add unit tests for logging util

awsarron commented on src/strands_agents_builder/utils/logging_utils.py:
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+"""
+Utility functions for configuring and managing logging in the Strands CLI.
```
nit: Strands Agent Builder*

awsarron commented on src/strands_agents_builder/utils/logging_utils.py:
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
```
think the shebang can be removed

awsarron commented on src/strands_agents_builder/strands.py:
```diff
@@ -69,8 +71,26 @@ def main():
+    parser.add_argument(
+        "--log-file",
+        type=str,
+        help="Path to log file. If not specified, logging is disabled",
```
could it perhaps be useful to output on stderr in some cases if log file is not defined?

perhaps default log-level to INFO only if log-file is passed and log-level is not passed? Otherwise if log-level and log-file are not passed, don't enable logging. If only log-level is passed, perhaps output logs to stderr as `logging` does by default.

awsarron commented on src/strands_agents_builder/strands.py:
```diff
@@ -69,8 +71,26 @@ def main():
+    # Configure logging
+    configure_logging(log_level=args.log_level, log_file=args.log_file)
+    if args.log_file:
+        logging.info(f"Strands CLI started with log level {args.log_level}")
```
should we be using the root logger `logging` here, or instead get a logger for `strands_agents_builder` ?

awsarron commented on src/strands_agents_builder/utils/logging_utils.py:
```diff
@@ -0,0 +1,135 @@
+    )
+
+    # Configure specific Strands loggers
+    loggers = ["strands", "strands.agent", "strands.models", "strands.tools", "strands_agents_builder"]
```
`strands.agent`, `strands.models`, `strands.tools` might be redundant if `strands` is configured.

we may also want to configure `strands_tools`.

awsarron commented on src/strands_agents_builder/utils/logging_utils.py:
```diff
@@ -0,0 +1,135 @@
+    Returns:
+        List of log level names
+    """
+    return ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
```
do we need this function? Seems redundant when `logging` is what actually defines the log levels

mkmeral replied:
This is for the CLI interface

awsarron commented on src/strands_agents_builder/utils/logging_utils.py:
```diff
@@ -0,0 +1,135 @@
+    return ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"]
+
+
+def get_logging_status() -> Dict[str, Any]:
```
this function isn't used

awsarron commented on src/strands_agents_builder/utils/logging_utils.py:
```diff
@@ -0,0 +1,135 @@
+    return status
+
+
+def set_log_level_for_module(module_name: str, log_level: Union[str, int]) -> None:
```
this function isn't used

Commit:
Merge branch 'main' into logging-utils

Commit:
Address all PR review comments for logging utilities

✅ Fixed Issues:
- Updated documentation from 'Strands CLI' to 'Strands Agent Builder'
- Removed shebang from utility module
- Enhanced logging behavior: stderr support when only --log-level provided
- Fixed logger usage: using module logger for startup messages
- Optimized logger configuration: simplified to parent loggers only
- Removed unused functions: get_logging_status() and set_log_level_for_module()

🚀 Improvements:
- Better UX: --log-level DEBUG works without --log-file (logs to stderr)
- Fallback handling: auto-fallback to stderr if file creation fails
- Enhanced CLI args: improved help text and argument behavior
- Comprehensive test suite: 9 unit tests covering all scenarios

📝 New Logging Behavior:
- No args = no logging (clean output)
- --log-level only = logs to stderr
- --log-file only = logs to file (INFO level)
- Both args = logs to file with specified level

All tests pass ✅

Commit:
fix(logging): remove logging levels function

Commit:
Merge branch 'main' into logging-utils
